libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,348 @@
+            {
+                return iter->second;
+            }
+            auto const node_id = m_graph.add_node(std::move(node));
+            m_solv2node[solv_id] = node_id;
+            return node_id;
+        };
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id,
+                                                   PackageInfo&& pkg_info,
+                                                   std::optional<ProblemType> pb_type) -> node_id
+        {
+            return ensure_solvable(solv_id, PackageNode{ std::move(pkg_info), pb_type });
+        }
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id,
+                                                   std::string&& dep,
+                                                   std::optional<ProblemType> pb_type) -> node_id
+        {
+            return ensure_solvable(solv_id, UnresolvedDependencyNode{ std::move(dep), pb_type });
+        }
+
+        auto ProblemsGraphCreator::add_conflict(node_id n1, node_id n2) -> void
+        {
+            m_conflicts[n1].insert(n2);
+            m_conflicts[n2].insert(n2);
+        }
+
+        auto ProblemsGraphCreator::add_expanded_deps_edges(node_id from_id,
+                                                           SolvId dep_id,
+                                                           edge_t const& edge) -> void
+        {
+            for (const auto& solv_id : m_pool.select_solvables(dep_id))
+            {
+                auto pkg_info = PackageInfo(pool_id2solvable(m_pool, solv_id));
+                auto to_id = ensure_solvable(solv_id, std::move(pkg_info), {});
+                m_graph.add_edge(from_id, to_id, edge);
+            }
+        }
+
+        auto ProblemsGraphCreator::parse_problems() -> void
+        {
+            for (const auto& problem : m_solver.all_problems_structured())
+            {
+                // These functions do not return a reference so we make sure to
+                // compute the value only once.
+                // TODO change name of these functions to make explicit it is not a ref.
+                auto source = problem.source();
+                auto target = problem.target();
+                auto dep = problem.dep();
+                auto type = map_problem(problem.type);
+                switch (problem.type)
+                {
+                    case SOLVER_RULE_PKG_CONSTRAINS:
+                    {
+                        // TODO
+                        if (!source || !target || !dep)
+                        {
+                            warn_unexpected_problem(problem);
+                            break;
+                        }
+                        auto src_id
+                            = ensure_solvable(problem.source_id, std::move(source).value(), type);
+                        auto tgt_id
+                            = ensure_solvable(problem.target_id, std::move(target).value(), type);
```

Comment:
Even though the type would be optional in this case - from https://github.com/syslaila/mamba-fork/pull/3/files#diff-e214b9fbd2b0497c09f4b5738a8aec7bdb8f107f5c7e3415be65d12898fc91f3R234. I wouldn't pass it here. 
I would expect the problem_type to be different for the source package & the target package. Same in the other cases as well.